### ISSUE TITLE
P2P sanity test failing

### ISSUE THREAD

**Issue by sethrosetter** [couchbase/couchbase-lite-java-core#1564]
Using:
  Android 1.4-26
  Mac 1.4-29

Failing test - http://uberjenkins.sc.couchbase.com/job/android-listener-client-client/19/testReport/testsuites.listener.shared.client_client/test_peer_2_peer_sanity/test_peer_2_peer_sanity/

```
V/Listener( 5373): CountDownLatch.await() END
D/dalvikvm( 5373): GC_FOR_ALLOC freed 435K, 11% free 7966K/8903K, paused 3ms, total 3ms
W/ChangeTracker( 5373): ChangeTracker{http://10.17.1.40:59850/ls_db2, OneShot, @535aed34}: Exception in change tracker
W/ChangeTracker( 5373): java.net.SocketTimeoutException: timeout
W/ChangeTracker( 5373): 	at okio.Okio$3.newTimeoutException(Okio.java:212)
W/ChangeTracker( 5373): 	at okio.AsyncTimeout.exit(AsyncTimeout.java:288)
W/ChangeTracker( 5373): 	at okio.AsyncTimeout$2.read(AsyncTimeout.java:242)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:325)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:314)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:210)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.Http1xStream.readResponse(Http1xStream.java:184)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.Http1xStream.readResponseHeaders(Http1xStream.java:125)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.readNetworkResponse(HttpEngine.java:775)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.access$200(HttpEngine.java:86)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:760)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.readResponse(HttpEngine.java:613)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.getResponse(RealCall.java:244)
W/ChangeTracker( 5373): 	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:201)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:163)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.execute(RealCall.java:57)
W/ChangeTracker( 5373): 	at com.couchbase.lite.replicator.ChangeTracker.runLoop(ChangeTracker.java:336)
W/ChangeTracker( 5373): 	at com.couchbase.lite.replicator.ChangeTracker.run(ChangeTracker.java:246)
W/ChangeTracker( 5373): 	at java.lang.Thread.run(Thread.java:856)
W/ChangeTracker( 5373): Caused by: java.net.SocketException: Socket closed
W/ChangeTracker( 5373): 	at libcore.io.Posix.recvfromBytes(Native Method)
W/ChangeTracker( 5373): 	at libcore.io.Posix.recvfrom(Posix.java:131)
W/ChangeTracker( 5373): 	at libcore.io.BlockGuardOs.recvfrom(BlockGuardOs.java:164)
W/ChangeTracker( 5373): 	at libcore.io.IoBridge.recvfrom(IoBridge.java:513)
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl.read(PlainSocketImpl.java:488)
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl.access$000(PlainSocketImpl.java:46)
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl$PlainSocketInputStream.read(PlainSocketImpl.java:240)
W/ChangeTracker( 5373): 	at okio.Okio$2.read(Okio.java:140)
W/ChangeTracker( 5373): 	at okio.AsyncTimeout$2.read(AsyncTimeout.java:238)
W/ChangeTracker( 5373): 	... 16 more
V/ChangeTracker( 5373): ChangeTracker{http://10.17.1.40:59850/ls_db2, OneShot, @535aed34}: Making request to POST [http://10.17.1.40:59850/ls_db2/_changes?feed=longpoll&heartbeat=30000&style=all_docs&active_only=true]
D/Listener( 5373): authHeader is null
V/Listener( 5373): Not enforcing basic auth -- allowedCredentials null or empty
V/Listener( 5373): remoteURL() addr=10.17.1.40
V/Listener( 5373): RouterCallbackBlock.onResponseReady() START
V/Listener( 5373): RouterCallbackBlock.onResponseReady() END
V/Listener( 5373): CountDownLatch.await() START
V/Listener( 5373): CountDownLatch.await() END
D/dalvikvm( 5373): GC_CONCURRENT freed 324K, 10% free 8030K/8903K, paused 11ms+0ms, total 15ms
W/ChangeTracker( 5373): ChangeTracker{http://10.17.1.40:59850/ls_db2, OneShot, @535aed34}: Exception in change tracker
W/ChangeTracker( 5373): java.net.SocketTimeoutException
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl.read(PlainSocketImpl.java:491)
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl.access$000(PlainSocketImpl.java:46)
W/ChangeTracker( 5373): 	at java.net.PlainSocketImpl$PlainSocketInputStream.read(PlainSocketImpl.java:240)
W/ChangeTracker( 5373): 	at okio.Okio$2.read(Okio.java:140)
W/ChangeTracker( 5373): 	at okio.AsyncTimeout$2.read(AsyncTimeout.java:238)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:325)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:314)
W/ChangeTracker( 5373): 	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:210)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.Http1xStream.readResponse(Http1xStream.java:184)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.Http1xStream.readResponseHeaders(Http1xStream.java:125)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.readNetworkResponse(HttpEngine.java:775)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.access$200(HttpEngine.java:86)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:760)
W/ChangeTracker( 5373): 	at okhttp3.internal.http.HttpEngine.readResponse(HttpEngine.java:613)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.getResponse(RealCall.java:244)
W/ChangeTracker( 5373): 	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:201)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:163)
W/ChangeTracker( 5373): 	at okhttp3.RealCall.execute(RealCall.java:57)
W/ChangeTracker( 5373): 	at com.couchbase.lite.replicator.ChangeTracker.runLoop(ChangeTracker.java:336)
W/ChangeTracker( 5373): 	at com.couchbase.lite.replicator.ChangeTracker.run(ChangeTracker.java:246)
W/ChangeTracker( 5373): 	at java.lang.Thread.run(Thread.java:856)
V/ChangeTracker( 5373): ChangeTracker{http://10.17.1.40:59850/ls_db2, OneShot, @535aed34}: Making request to POST [http://10.17.1.40:59850/ls_db2/_changes?feed=longpoll&heartbeat=30000&style=all_docs&active_only=true]
```

**Comment by hideki** [comment]
@sethrosetter:
The test fails because `macosx`'s  pull replication status never becomes `IDLE`. I assume this is CBL iOS issue. There is possibility that CBL Android returns the unexpected return value to CBL iOS.

Error message from pytest:
```
        except _ProxyError as e:
            raise ProxyError(e)
    
        except (_SSLError, _HTTPError) as e:
            if isinstance(e, _SSLError):
                raise SSLError(e, request=request)
            elif isinstance(e, ReadTimeoutError):
>               raise ReadTimeout(e, request=request)
E               ReadTimeout: HTTPConnectionPool(host='10.17.1.60', port=51000): Read timed out. (read timeout=None)

venv/lib/python2.7/site-packages/requests/adapters.py:449: ReadTimeout

Tearing down test
```

android - http://192.168.57.101:5000/_active_tasks
```json
[
{
progress: 0,
continuous: true,
source: "http://10.17.1.60:51000/ls_db2",
status: "Idle",
task: "repl002",
change_count: 0,
target: "ls_db1",
completed_change_count: 0,
type: "Replication"
},
{
progress: 0,
continuous: true,
source: "ls_db1",
status: "Idle",
task: "repl001",
change_count: 0,
target: "http://10.17.1.60:51000/ls_db2",
completed_change_count: 0,
type: "Replication"
}
]
```
macosx - http://10.17.1.60:51000/_active_tasks
```json
[
{
continuous: true,
status: "Idle",
task: "repl001",
type: "Replication",
source: "ls_db2",
target: "http://192.168.57.101:5000/ls_db1"
},
{
continuous: true,
status: "Processed 0 / 0 changes",
task: "repl002",
type: "Replication",
source: "http://192.168.57.101:5000/ls_db1",
target: "ls_db2"
}
]
```

**Comment by hideki** [comment]
Based on my investigation, I believe this is CBL iOS side issue. I will move this ticket into Review column.

**Comment by pasin** [comment]
Does iOS to iOS or Android to Android fail?

**Comment by pasin** [comment]
It would be possible that the iOS/MacOS pull replicator doesn't go to idle because it's still waiting for  data from the android?

**Comment by hideki** [comment]
- osx to osx is not testable. 
- to test android to android, requires to update test scripts. With two androids, adb command is confused by multiple emulator connections.

**Comment by sethrosetter** [comment]
I tested osx to osx. This works as expected.

**Comment by sethrosetter** [comment]
Blocked on testing Android to Android due to - https://github.com/couchbase/couchbase-lite-java-core/issues/1563

**Comment by hideki** [comment]
Notes:
- now osx to osx works fine with my test env.
- From this result, it seems CBL Android might return the JSON response CBL iOS can not understand or CBL Android does not respond for CBL iOS request.

**Comment by hideki** [comment]
It seems LiteServ Android blocks response for `POST /_changes` 
```
hideki@hidekis:~$ curl -X POST "http://192.168.57.101:5000/ls_db1/_changes?feed=normal&heartbeat=300000&style=all_docs&active_only=true"
{"error":"request_timeout","status":408}hideki@hidekis:~$
```

**Comment by hideki** [comment]
Note to Self:

- CBL iOS's replicator does not become IDLE state because CBL iOS does not receive response POST /_changes REST API from CBL Android.
- When CBL Android tries to send JSON response, connection with CBL iOS is already closed. Then throws Exception. 

Currently, I don't know why the connection is closed????

**Comment by hideki** [comment]
Hi @sethrosetter and @pasin 

I tested following combinations:
- android 1.4 <-> net-mono (1.4-29)
- android 1.4 <-> macosx (1.3.1-16)
Both combinations can pass the test.

Based on my investigation, connection from macosx for `POST /_changes` REST API is closed when android tries to send JSON response. Therefore, macosx can not receive the response for `POST /_changes`, this is reason pull replication status is not changed to IDLE.

From Android side point of view, I am no idea why the connection is closed.

Can you look into CBL iOS side? This test can pass with CBL iOS 1.3.1-16. There is possibility that network related code change in iOS codes could affect for this.

Thanks!

**Comment by hideki** [comment]
@pasin 

LiteServ.iOS error log for `/_changes`
```
14:51:21.830‖ ChangeTracker: CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: Starting...
14:51:21.830‖ Sync: CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: POST //192.168.57.101:5000/ls_db1/_changes?feed=normal&heartbeat=300000&style=all_docs&active_only=true
14:51:21.830‖ ChangeTracker: CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: POST //192.168.57.101:5000/ls_db1/_changes?feed=normal&heartbeat=300000&style=all_docs&active_only=true {
    "Accept-Encoding" = gzip;
    "Content-Length" = 99;
    "Content-Type" = "application/json";
    Host = "192.168.57.101:5000";
    "User-Agent" = "CouchbaseLite/1.3 (Mac OS X 1.4/e56330cb365853cfd146be620b5f854f3ab7afae)";
}
14:51:21.831‖ ChangeTracker: Changes feed using proxy settings {
    ExceptionsList =     (
        "*.local",
        "169.254/16"
    );
    FTPPassive = 1;
    "__SCOPED__" =     {
        en0 =         {
            ExceptionsList =             (
                "*.local",
                "169.254/16"
            );
            FTPPassive = 1;
        };
    };
}
14:51:21.831‖ ChangeTracker: CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: Started... <http://192.168.57.101:5000/ls_db1/_changes?feed=normal&heartbeat=300000&style=all_docs&active_only=true>
14:51:21.831‖ ChangeTracker: CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: Event 1 on <__NSCFInputStream: 0x7fcab521ff50>
14:51:21.841‖ CBLSocketChangeTracker[0x7fcab362b970 ls_db1]: Connection error #2, retrying in 4.0 sec: CFCFNetworkError[-1005]
14:51:21.890‖ Router: GET http://10.17.1.60:51000/_active_tasks
	Accept-Encoding: identity
	Host: 10.17.1.60:51000
	Content-Type: application/json
14:51:21.890‖ Router: Response -- status=200, body=301 bytes
	Cache-Control: must-revalidate
	Server: CouchbaseLite 1.4 (build 29)
	Content-Type: application/json
```

**Comment by hideki** [comment]
Duplicate/Reference: https://github.com/couchbase/couchbase-lite-java-core/issues/1470

**Comment by hideki** [comment]
## Notes
#### Problem description
1. CBL iOS calls REST API with "connection: closed" header value.
2. In case of having the request body in HTTP request, CBL Android Router parses the request body with using Jackson.
3. Jackson parser automatically closes `InputStream` after reading data from InputStream. Actually call `InputStream.close()`
4. InputStream implementation is in TJWS codes, and its instance is created/managed by TJWS.
5. TJWS closes the connection in InputStream.close() implementation if it is  keepalive!= true (connection: close).

#### TJWS `InputStream.close()` implementation
https://github.com/couchbase/couchbase-lite-java-listener/blob/master/vendor/tjws/src/java/Acme/Serve/Serve.java#L4546-L4573
When InputStream from Socket is closed, connection is closed. <- observation from behavior

#### Jackson Auto closing
- `AUTO_CLOSE_SOURCE ` - https://github.com/FasterXML/jackson-core/wiki/JsonParser-Features
- https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/ObjectMapper.java#L2955-L2960

**Comment by hideki** [comment]
Solution:
- Disable Jackson's AUTO_CLOSE_SOURCE feature
- Make sure all InputStream are closed after used.

**Comment by hideki** [comment]
Issue will be fixed with build 1.4-28
http://latestbuilds.hq.couchbase.com/couchbase-lite-android/1.4/1.4-28/
